src/components/general/AdaptiveCardViewer/index.tsx

```diff
@@ -0,0 +1,86 @@
+    onExecuteAction?: (action: AdaptiveCards.Action) => void;
+    /** Method that will be invoked when a Submit action is executed. [More Info](https://docs.microsoft.com/en-us/adaptive-cards/display/implementingrenderer#actionsubmit) */
+    onActionSubmit?: (action: AdaptiveCards.Action) => void;
+    /** Method that will be invoked when an Open Url action is executed. [More Info](https://docs.microsoft.com/en-us/adaptive-cards/display/implementingrenderer#actionopenurl) */
+    onActionOpenUrl?: (action: AdaptiveCards.Action) => void;
+    /** Method that will be invoked when a Show Card action is executed. [More Info](https://docs.microsoft.com/en-us/adaptive-cards/display/implementingrenderer#actionshowcard) */
+    onActionShowCard?: (action: AdaptiveCards.Action) => void;
+    /** CSS classes that will bew applied to the card container */
+    className?: string;
+};
+
+const ACTION_OPEN_URL = 'Action.OpenUrl';
+const ACTION_SHOW_CARD = 'Action.ShowCard';
+const ACTION_SUBMIT = 'Action.Submit';
+
+const AdaptiveCardViewer: React.FC<AdaptiveCardViewerProps> = ({
+    hostConfig,
+    payload,
+    onExecuteAction,
+    onActionOpenUrl,
+    onActionShowCard,
+    onActionSubmit,
+    className,
+}) => {
+    const cardContainerRef = React.useRef<HTMLDivElement | null>(null);
+    const adaptiveCard = React.useMemo(() => new AdaptiveCards.AdaptiveCard(), []);
+
+    AdaptiveCards.AdaptiveCard.onProcessMarkdown = (text, result) => {
+        result.outputHtml = marked(text);
+        result.didProcess = true;
+    };
+
+    if (hostConfig) {
+        adaptiveCard.hostConfig = new AdaptiveCards.HostConfig(hostConfig);
+    }
+
+    const executeAction = React.useCallback(
+        (action: AdaptiveCards.Action) => {
+            switch (action.getJsonTypeName()) {
+                case ACTION_OPEN_URL:
+                    onActionOpenUrl
+                        ? onActionOpenUrl(action)
+                        : window.open(action.getHref(), action.title || '_blank');
+                    break;
+                case ACTION_SHOW_CARD:
+                    onActionShowCard && onActionShowCard(action);
+                    break;
+                case ACTION_SUBMIT:
+                    onActionSubmit && onActionSubmit(action);
+                    break;
+            }
+            onExecuteAction && onExecuteAction(action);
+        },
+        [onExecuteAction, onActionOpenUrl, onActionShowCard, onActionSubmit]
+    );
+
+    adaptiveCard.onExecuteAction = executeAction;
+
+    const result = React.useMemo(() => {
+        adaptiveCard.parse(payload);
+        return adaptiveCard.render();
+    }, [payload]);
+
+
+    React.useEffect(() => {
+        if (cardContainerRef && cardContainerRef.current) {
+            const current = cardContainerRef.current;
+            current.firstChild && current.removeChild(current.firstChild);
+            current.appendChild(result);
```

Comment:
Instead of removing the child _before_ appending the new one, we should remove the child in the cleanup method for the effect. That would make the entire process "async" instead of blocking

Reply:
Now removes child in useEffect cleanup function